octodns/provider/gandi.py

```diff
@@ -0,0 +1,334 @@
+#
+#
+#
+
+from __future__ import absolute_import, division, print_function, \
+    unicode_literals
+
+from collections import defaultdict
+from requests import Session
+import logging
+
+from ..record import Record
+from .base import BaseProvider
+
+
+class GandiClientException(Exception):
+    pass
+
+
+class GandiClientBadRequest(GandiClientException):
+
+    def __init__(self, r):
+        super(GandiClientBadRequest, self).__init__(r.text)
+
+
+class GandiClientUnauthorized(GandiClientException):
+
+    def __init__(self, r):
+        super(GandiClientUnauthorized, self).__init__(r.text)
+
+
+class GandiClientForbidden(GandiClientException):
+
+    def __init__(self, r):
+        super(GandiClientForbidden, self).__init__(r.text)
+
+
+class GandiClientNotFound(GandiClientException):
+
+    def __init__(self, r):
+        super(GandiClientNotFound, self).__init__(r.text)
+
+
+class GandiClient(object):
+
+    def __init__(self, token):
+        session = Session()
+        session.headers.update({'Authorization': 'Apikey {}'.format(token)})
+        self._session = session
+        self.endpoint = 'https://api.gandi.net/v5'
+
+    def _request(self, method, path, params={}, data=None):
+        url = '{}{}'.format(self.endpoint, path)
+        r = self._session.request(method, url, params=params, json=data)
+        if r.status_code == 400:
+            raise GandiClientBadRequest(r)
+        if r.status_code == 401:
+            raise GandiClientUnauthorized(r)
+        elif r.status_code == 403:
+            raise GandiClientForbidden(r)
+        elif r.status_code == 404:
+            raise GandiClientNotFound(r)
+        r.raise_for_status()
+        return r
+
+    def zone_records(self, zone_name):
+        records = self._request('GET', '/livedns/domains/{}/records'
+                                .format(zone_name)).json()
+
+        for record in records:
+            if record['rrset_name'] == '@':
+                record['rrset_name'] = ''
+
+        return records
+
+    def record_create(self, zone_name, data):
+        self._request('POST', '/livedns/domains/{}/records'.format(zone_name),
+                      data=data)
+
+    def record_delete(self, zone_name, record_name, record_type):
+        self._request('DELETE', '/livedns/domains/{}/records/{}/{}'
+                      .format(zone_name, record_name, record_type))
+
+
+class GandiProvider(BaseProvider):
+    '''
+    Gandi provider using API v5.
+
+    gandi:
+        class: octodns.provider.gandi.GandiProvider
+        # Your API key (required)
+        token: XXXXXXXXXXXX
+    '''
+
+    SUPPORTS_GEO = False
+    SUPPORTS_DYNAMIC = False
+    SUPPORTS = set((['A', 'AAAA', 'ALIAS', 'CAA', 'CNAME', 'DNAME',
```

Comment:
I don't believe DNAME is used by any other provider, nor does it have validation code to be handled. I believe it can be removed from this supported list to avoid issues.

Reply:
I just submitted a PR to add support for DNAME records (#620).

Reply:
Merged.